File: 백준/js/boj_10819.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = require("fs").readFileSync(filePath).toString().trim().split("\n");

const N = Number(input[0]);
input = input[1].split(" ").map(Number);

const check = new Array(N).fill(false);
const newArr = [];
let max = 0;

function cal(arr) {
  let sum = 0;
  for (let i = 0; i < N - 1; i++) {
    sum += Math.abs(arr[i] - arr[i + 1]);
  }
  return sum;
}

function dfs(depth) {
  for (let i = 0; i < N; i++) {
    if (depth === N) {
      max = Math.max(max, cal(newArr));
    }

    if (!check[i]) {
      check[i] = true;
      newArr.push(input[i]);
      dfs(depth + 1);
      check[i] = false;
      newArr.pop();
    }
  }
}

dfs(0);
console.log(max);

File: 백준/python/boj_10819.py
import sys

sys.stdin = open("input.txt", "r")

N = int(input())
arr = list(map(int, input().split()))

# 백트래킹 풀이
check = [False] * N
newArr = []
ans = 0


def cal(array):
    total = 0
    for i in range(N - 1):
        total += abs(array[i] - array[i + 1])
    return total


def dfs(depth):
    global ans
    if depth == N:
        ans = max(ans, cal(newArr))

    for i in range(N):
        if not check[i]:
            check[i] = True
            newArr.append(arr[i])
            dfs(depth + 1)
            check[i] = False
            newArr.pop()


dfs(0)
print(ans)


# 순열 풀이
from itertools import permutations

answer = 0

def cal(array):
    total = 0
    for i in range(N - 1):
        total += abs(array[i] - array[i + 1])
    return total

for list in permutations(arr):
    answer = max(answer, cal(list))

print(answer)